platform/search/src/GenericSearchProvider.js

```diff
@@ -0,0 +1,259 @@
+                var ids = [],
+                    id;
+                
+                // If we have the results from a search 
+                if (event.data.request === 'search') {
+                    // Convert the ids given from the web worker into domain objects
+                    for (id in event.data.results) {
+                        ids.push(id);
+                    }
+                    objectService.getObjects(ids).then(function (objects) {
+                        var searchResults = [],
+                            id;
+                        
+                        // Create searchResult objects
+                        for (id in objects) {
+                            searchResults.push({
+                                object: objects[id],
+                                id: id,
+                                score: event.data.results[id]
+                            });
+                        }
+                        
+                        // Resove the promise corresponding to this 
+                        pendingQueries[event.data.timestamp].resolve({
+                            hits: searchResults,
+                            total: event.data.total,
+                            timedOut: event.data.timedOut
+                        });
+                    });
+                }
+            }
+            
+            worker.onmessage = handleResponse;
+            
+            // Helper function for getItems(). Indexes the tree.
+            function indexItems(nodes) {
+                nodes.forEach(function (node) {
+                    var id = node && node.getId && node.getId();
+                    
+                    // If we have not yet indexed this item, index it 
+                    if (!indexed[id]) {
+                        // Index each item with the web worker
+                        indexItem(node);
+                        indexed[id] = true;
+                        
+                        if (node.hasCapability && node.hasCapability('composition')) {
+                            // This node has children
+                            node.getCapability('composition').invoke().then(function (children) {
+                                // Index the children
+                                if (children.constructor === Array) {
+                                    indexItems(children);
+                                } else {
+                                    indexItems([children]);
+                                }
+                            });
+                        }
+                    }
+                    
+                    // Watch for changes to this item, in case it gets new children 
+                    if (node && node.hasCapability && node.hasCapability('mutation')) {
+                        node.getCapability('mutation').listen(function (listener) {
+                            if (listener && listener.composition) {
```

Comment:
I'd replace this function with:

```
indexed[node.getId()] = false; // Mark as un-indexed so we can reindex
indexItems([node]);            // ...and reindex
```

The children in this case may not need to be re-indexed (the check against the `indexed` set will help here)